tests/transforms/test_add_noise.py

```diff
@@ -316,6 +328,75 @@ def explicit_circuit(circuit_param):
 
         assert np.allclose(noisy_circuit(0.4), explicit_circuit(0.4))
 
+    # pylint:disable = cell-var-from-loop
+    def test_add_noise_with_readout_errors(self):
+        """Test that a add_noise works with readout errors."""
+        dev = qml.device("default.mixed", wires=2)
+
+        fc, fn = qml.noise.op_in([qml.RY, qml.RZ]), qml.noise.partial_wires(
+            qml.AmplitudeDamping, 0.4
+        )
+        mc, mn = (qml.noise.meas_eq(qml.expval) | qml.noise.meas_eq(qml.var)) & qml.noise.wires_in(
+            [0, 1]
+        ), qml.noise.partial_wires(qml.PhaseFlip, 0.2)
+
+        @partial(add_noise, noise_model=qml.NoiseModel({fc: fn}, {mc: mn}))
+        @qml.qnode(dev)
+        def f_noisy(w, x, y, z):
+            qml.RX(w, wires=0)
+            qml.RY(x, wires=1)
+            qml.CNOT(wires=[0, 1])
+            qml.RY(y, wires=0)
+            qml.RX(z, wires=1)
+            return (
+                qml.expval(qml.PauliZ(0) @ qml.PauliZ(1)),
+                qml.probs(op=qml.Z(0) @ qml.Z(1)),
+                qml.purity(wires=0),
+                qml.var(qml.PauliZ(0) @ qml.PauliZ(1)),
+            )
+
+        args = [0.1, 0.2, 0.3, 0.4]
+
+        results = []
+        for mp in [
+            qml.expval(qml.PauliZ(0) @ qml.PauliZ(1)),
+            qml.var(qml.PauliZ(0) @ qml.PauliZ(1)),
+        ]:
+
+            @qml.qnode(dev)
+            def f(w, x, y, z):
+                qml.RX(w, wires=0)
+                qml.RY(x, wires=1)
+                qml.AmplitudeDamping(0.4, wires=1)
+                qml.CNOT(wires=[0, 1])
+                qml.RY(y, wires=0)
+                qml.AmplitudeDamping(0.4, wires=0)
+                qml.RX(z, wires=1)
+                qml.PhaseFlip(0.2, wires=0)
+                qml.PhaseFlip(0.2, wires=1)
+                return qml.apply(mp)
+
+            results.append(f(*args))
+
+        for mp in [qml.probs(op=qml.Z(0) @ qml.Z(1)), qml.purity(wires=0)]:
+
+            @qml.qnode(dev)
+            def g(w, x, y, z):
+                qml.RX(w, wires=0)
+                qml.RY(x, wires=1)
+                qml.AmplitudeDamping(0.4, wires=1)
+                qml.CNOT(wires=[0, 1])
+                qml.RY(y, wires=0)
+                qml.AmplitudeDamping(0.4, wires=0)
+                qml.RX(z, wires=1)
+                return qml.apply(mp)
+
+            results.append(g(*args))
+
+        # assert not np.isclose(f_noisy(*args), f(*args))
+        for res1, res2 in zip(f_noisy(*args), (results[0], results[2], results[3], results[1])):
+            assert qml.math.allclose(res1, res2)
```

Comment:
Better to avoid for loops in assert, in case of failure it could be hard to see what is going wrong.